File: public/javascripts/sort/HeapSort.js
function HeapSort() {
  this.heap = []
};

HeapSort.prototype.sort = function(collection) {
  this.heap = collection;
  this.heapify();
  this.toSequence();
  return this.heap;
};

HeapSort.prototype.heapify = function() {
  var length = this.heap.length-1;

  for(var start = Math.floor((length-1)/2); start >= 0; start--) {
    this.sortDown(start, length);
  }
};

HeapSort.prototype.toSequence = function() {
  for(var last = this.heap.length-1; last > 0; last--) {
    this.swap(0, last);
    this.sortDown(0, last-1);
  }
};

HeapSort.prototype.swap = function(a, b) {
  var temp = this.heap[a];
  this.heap[a] = this.heap[b];
  this.heap[b] = temp;
};

HeapSort.prototype.sortDown = function(root, heapEnd) {
  while(2*root+1 <= heapEnd) {
    var leftChild = 2*root+1;
    var rightChild = leftChild+1;

    if(rightChild <= heapEnd &&
      this.heap[leftChild] < this.heap[rightChild] &&
      this.heap[root] < this.heap[rightChild]) {

      this.swap(root, rightChild);
      root = rightChild;

    } else if(this.heap[root] < this.heap[leftChild]) {

      this.swap(root, leftChild);
      root = leftChild;

    } else {
      return;
    }
  }
};

File: lib/sort/Heap_Sort.rb
class Heap_Sort

  def sort collection
    @heap = collection
    @size = collection.size-1
    heapify
    sequentialize
    @heap
  end

  private
  def sequentialize last = @size
    return if last <= 0
    swap 0, last
    sort_down 0, last-1
    sequentialize last-1
  end

  def heapify parent = @size/2
    return if parent < 0
    sort_down parent
    heapify parent-1
  end

  def swap parent, child
    @heap[parent], @heap[child] = @heap[child], @heap[parent]
  end

  def sort_down parent, heap_end = @size
    left_child = 2*parent + 1
    right_child = left_child + 1

    if right_child <= heap_end and
      @heap[right_child] > @heap[parent] and
      @heap[left_child] < @heap[right_child]

      swap parent, right_child
      sort_down right_child, heap_end

    elsif left_child <= heap_end and
      @heap[parent] < @heap[left_child]

      swap parent, left_child
      sort_down left_child, heap_end
    end
  end
end
